docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md

```diff
@@ -0,0 +1,433 @@
+
+For example, consider the following code:
+```python
+def calculate_total_income(gross_wages, taxable_interest, total_dividends, qualified_dividends, other_income):
+    total_income = gross_wages + taxable_interest + (total_dividends - qualified_dividends) + other_income
+    return total_income
+```
+
+It is clear that the function definition is too long to fit in one line, so it needs to be broken up. But where should the line breaks go?
+One option is to break up the long lines after each comma or operator:
+```python
+def calculate_total_income(gross_wages, 
+                           taxable_interest, 
+                           total_dividends, 
+                           qualified_dividends, 
+                           other_income):
+    total_income = (gross_wages
+                    + taxable_interest
+                    + (total_dividends - qualified_dividends)
+                    + other_income)
+    return total_income
+```
+
+Another option is to use a separate line for each argument and also for the closing parentheses:
+```python
+def calculate_total_income(
+    gross_wages,
+    taxable_interest, 
+    total_dividends, 
+    qualified_dividends, 
+    other_income
+):
+    total_income = (
+        gross_wages 
+        + taxable_interest 
+        + (total_dividends - qualified_dividends) 
+        + other_income
+    )
+    return total_income
+```
+Although these two versions are syntactically identical and formatted similarly, the formatting rules they obey are quite different. Even worse, a "diff" or line-by-line comparison of the two versions is very hard to read, since almost every line is different in some trivial way. This means that if a programmer both reformatted the code from one format to the other and, at the same time, made a substantive change to it (like renaming one of the arguments of the `calculate_total_income` function), it would be hard to quickly determine which lines were meaningfully changed and which were just reformatted.
+
+### Code formatters enforce formatting conventions
+To both eliminate these kinds of formatting ambiguities and to automate the otherwise-tedious process of formatting code by hand, tools to automatically format code have been developed. These tools take files of source code as input and output the same code, but reformatted as necessary according to a pre-specified set of formatting rules.
+
+In Python, the most popular code formatting tool is called `black`. This tool both automatically formats code and imposes its own opinionated set of formatting rules (hence its tagline: "any color you want"). This eliminates the need for developers to make (or argue about) these decisions themselves. Although `black`'s formatting rules can take some getting used to, the benefit of eliminating almost all formatting-related decisions is usually worth the effort of adjusting to its opinions. 
+
+
+### Setting up `black`
```

Comment:
Since we use ruff for linting and ruff is meant to be a drop in replacement for black, what is the reason not to cover this section with ruff? Educational? If so, I'd call out the fact that ruff can be used instead. 

And in the case of Jupyter notebooks, Ruff has native support so you wouldn't need nb-black etc.

Reply:
I was also a little confused by this; is it b/c we want an example of a formatter that is separate from a linter?

Reply:
I used `black` in the lesson because I thought it would be clearer and I wanted to give credit to `black` for its role in standardizing formatting in python. The fact that `ruff format` reimplements black and is faster feels like an implementation detail. But I will add an explicit mention of this at the end when we mention the template repos.

Reply:
Yeah, but we don't people to get to sense that they need both. So good to clarify